File: Javascript/Two Pointers/M-443. String Compression.ts
function compress(chars: string[]): number {
  let [l, r] = [0, 0];
  while (r < chars.length) {
    const curChar = chars[r];
    let count = 0;

    /* find the end position of curChar */
    while (r < chars.length && curChar === chars[r]) {
      r++;
      count++;
    }

    /* add curChar */
    chars[l++] = curChar;

    /* add count */
    if (count > 1) {
      for (const digit of count.toString()) chars[l++] = digit;
    }
  }

  return l;
}

const testData = [
  {
    chars: ['a', 'a', 'b', 'b', 'c', 'c', 'c'],
    ans: 6,
  },
  {
    chars: ['a'],
    ans: 1,
  },
  {
    chars: ['a', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b'],
    ans: 4,
  },
];

for (const data of testData) {
  console.log(data['ans'] === compress(data['chars']));
}

File: Python/Two Pointers/M-443. String Compression.py
class Solution:
    @classmethod
    def compress(self, chars: list[str]) -> int:
        l, r = 0, 0

        while r < len(chars):
            cur_char = chars[r]
            count = 0

            # find the end position of char_char
            while r < len(chars) and cur_char == chars[r]:
                r += 1
                count += 1

            # add current char
            chars[l] = cur_char
            l += 1

            # add count
            if count > 1:
                for digit in str(count):
                    chars[l] = digit
                    l += 1

        return l


testData = [
    {
        "chars": ["a", "a", "b", "b", "c", "c", "c"],
        "ans": 6,
    },
    {
        "chars": ["a"],
        "ans": 1,
    },
    {
        "chars": ["a", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b"],
        "ans": 4,
    },
]

for data in testData:
    print(data["ans"] == Solution.compress(data["chars"]))
